p2p/src/p2p_reducer.rs

```diff
@@ -1,149 +1,70 @@
-                                    })
-                                }
-                            };
-                            P2pConnectionOutgoingInitOpts::WebRTC {
-                                peer_id: *peer_id,
-                                signaling,
-                            }
-                        }),
-                        status: P2pPeerStatus::Connecting(P2pConnectionState::incoming_init(opts)),
-                        identify: None,
-                    }),
-                    P2pConnectionAction::Incoming(
-                        P2pConnectionIncomingAction::FinalizePendingLibp2p { .. },
-                    ) => {
-                        state.peers.entry(*peer_id).or_insert_with(|| P2pPeerState {
-                            is_libp2p: true,
-                            dial_opts: None,
-                            // correct status later set in the child reducer.
-                            status: P2pPeerStatus::Disconnected { time: meta.time() },
-                            identify: None,
-                        })
-                    }
-                    _ => match state.peers.get_mut(peer_id) {
-                        Some(v) => v,
-                        None => return,
-                    },
-                };
-                p2p_connection_reducer(peer, my_id, meta.with_action(action));
+                P2pConnectionState::reducer(state_context, meta.with_action(action))
             }
-            P2pAction::Disconnection(action) => match action {
-                P2pDisconnectionAction::Init { .. } => {}
-                P2pDisconnectionAction::Finish { peer_id } => {
-                    #[cfg(feature = "p2p-libp2p")]
-                    if state
-                        .network
-                        .scheduler
-                        .connections
-                        .iter()
-                        .any(|(_addr, conn_state)| conn_state.peer_id() == Some(peer_id))
-                    {
-                        // still have other connections
-                        return;
-                    }
-                    let Some(peer) = state.peers.get_mut(peer_id) else {
-                        return;
-                    };
-                    peer.status = P2pPeerStatus::Disconnected { time: meta.time() };
-                }
-            },
-            P2pAction::Peer(action) => {
-                p2p_peer_reducer(state, meta.with_action(action));
+            P2pAction::Disconnection(action) => {
+                P2pDisconnectedState::reducer(state_context, meta.with_action(action))
             }
+            P2pAction::Peer(action) => P2pPeerState::reducer(
+                Substate::from_compatible_substate(state_context),
+                meta.with_action(action),
+            ),
             P2pAction::Channels(action) => {
                 let Some(peer_id) = action.peer_id() else {
-                    return;
+                    return Ok(());
```

Comment:
I assume these are not `bug_condition`s and expected

Reply:
Yes, there was one like this for connections but I updated and removed it